lib/lambda_ethereum_consensus/state_transition/predicates.ex

```diff
@@ -102,6 +102,35 @@ defmodule LambdaEthereumConsensus.StateTransition.Predicates do
     is_indices_available(validators, indices, h < validators)
   end
 
+  @doc """
+  Check if merkle branch is valid
+  """
+  @spec is_valid_merkle_branch(
+          SszTypes.bytes32(),
+          list(SszTypes.bytes32()),
+          SszTypes.uint64(),
+          SszTypes.uint64(),
+          SszTypes.root()
+        ) :: boolean
+  def is_valid_merkle_branch(leaf, branch, depth, index, root) do
+    is_valid_merkle_branch(leaf, branch, depth, index, root, leaf, 0)
+  end
+
+  defp is_valid_merkle_branch(leaf, branch, depth, index, root, value, i) do
+    if i == depth do
+      IO.inspect(value == root)
+      value == root
+    else
+      if rem(div(index, 2 ** i), 2) == 1 do
+        value = :crypto.hash(:sha256, Enum.at(branch, i) <> value)
+        is_valid_merkle_branch(leaf, branch, depth, index, root, value, i + 1)
+      else
+        value = :crypto.hash(:sha256, value <> Enum.at(branch, i))
+        is_valid_merkle_branch(leaf, branch, depth, index, root, value, i + 1)
+      end
+    end
+  end
```

Comment:
Let's change this implementation to the following structure:

```elixir
def is_valid_merkle_branch?(leaf, branch, depth, index, root) do
  root == branch
    |> Enum.take(depth)
    |> Enum.with_index()
    |> Enum.reduce(leaf, fn {i, v}, value -> hash_merkle_node(v, value, index, i) end)
end

defp hash_merkle_node(value_1, value_2, index, i) when rem(div(index, 2 ** i), 2) == 1 do
  :crypto.hash(:sha256, value_1 <> value_2)
end

defp hash_merkle_node(value_1, value_2, _index, _i) do
  :crypto.hash(:sha256, value_2 <> value_1)
end
```

Aside from readability, this prevents us from using `Enum.at()` in each iteration, so we go from quadratic to linear complexity.

Reply:
There is an issue about using rem and div in when guard;

error: cannot invoke remote function Kernel.**/2 inside guards

Reply:
I still think it's better to have this in a reduce, without the manual recursion. If we can't use the guards, we can just use an `if` for the `hash_merkle_node` function:

```elixir
defp hash_merkle_node(value_1, value_2, index, i) do
  if rem(div(index, 2 ** i), 2) == 1 do
    :crypto.hash(:sha256, value_1 <> value_2)
  else
    :crypto.hash(:sha256, value_2 <> value_1)
  end
end
```